okapi-core/src/main/java/org/folio/okapi/service/impl/ProcessModuleHandle.java

```diff
@@ -175,100 +175,81 @@ private void start2(Handler<AsyncResult<Void>> startFuture) {
         }
       }
       future.complete();
-    }, false, result -> {
-      if (result.failed()) {
-        logger.debug("ProcessModuleHandle.start2() executeBlocking failed {}",
-            result.cause().getMessage());
-        startFuture.handle(Future.failedFuture(result.cause()));
-      } else {
-        start3(startFuture);
-      }
-    });
+    }, false).compose(x -> start3());
   }
 
-  private void start3(Handler<AsyncResult<Void>> startFuture) {
-    tcpPortWaiting.waitReady(process, x -> {
-      if (x.failed()) {
-        this.stopProcess(y -> startFuture.handle(Future.failedFuture(x.cause())));
-      } else {
-        startFuture.handle(Future.succeededFuture());
-      }
+  private Future<Void> start3() {
+    return tcpPortWaiting.waitReady(process).onFailure(x -> {
+      this.stopProcess();
     });
   }
 
-  private void waitPortToClose(Handler<AsyncResult<Void>> stopFuture, int iter) {
-    if (port > 0) {
-      // fail if port is already in use
-      NetClientOptions options = new NetClientOptions().setConnectTimeout(50);
-      NetClient c = vertx.createNetClient(options);
-      c.connect(port, "localhost", res -> {
-        if (res.succeeded()) {
-          NetSocket socket = res.result();
-          socket.close();
-          if (iter > 0) {
-            vertx.setTimer(100, x -> waitPortToClose(stopFuture, iter - 1));
-          } else {
-            stopFuture.handle(Future.failedFuture(
-                messages.getMessage("11503", Integer.toString(port))));
-          }
+  private Future<Void> waitPortToClose(int iter) {
+    if (port == 0) {
+      return Future.succeededFuture();
+    }
+    NetClientOptions options = new NetClientOptions().setConnectTimeout(50);
+    NetClient c = vertx.createNetClient(options);
+    Promise<Void> promise = Promise.promise();
+    c.connect(port, "localhost").onComplete(res -> {
+      if (res.succeeded()) {
+        NetSocket socket = res.result();
+        socket.close();
+        if (iter > 0) {
+          vertx.setTimer(100, id -> waitPortToClose(iter - 1).onComplete(promise::handle));
         } else {
-          stopFuture.handle(Future.succeededFuture());
+          promise.fail(messages.getMessage("11503", Integer.toString(port)));
         }
-      });
-    } else {
-      stopFuture.handle(Future.succeededFuture());
-    }
+      } else {
+        promise.complete();
+      }
+    });
+    return promise.future();
```

Comment:
Using compose results in more simple code:

    NetClientOptions options = new NetClientOptions().setConnectTimeout(50);
    NetClient c = vertx.createNetClient(options);
    return c.connect(port, "localhost").compose(socket -> {
      socket.close();
      if (iter > 0) {
        Promise<Void> promise = Promise.promise();
        vertx.setTimer(100, id -> waitPortToClose(iter - 1).onComplete(promise::handle));
        return promise.future();
      } else {
        return Future.failedFuture(messages.getMessage("11503", Integer.toString(port)));
      }
    }, fail -> Future.succeededFuture());

Reply:
yep